### ISSUE TITLE
Support non-linear font scaling

### ISSUE THREAD

**Issue by odbol** [robolectric/robolectric#8413]
**Is your feature request related to a problem? Please describe.**
Android SDK 34 supports non-linear font scaling by default. Robolectric should try to emulate it in some way.

**Describe the solution you'd like**

When loading a layout to be tested, any font sizes or dimensions in SP should be converted to pixels using the new non-linear font scaling tables. This only matters if the system font scale is > 1. 

The conversion should happen when the resource is loaded from XML. This typically uses the TypedValue.applyDimension() method to convert SP -> PX. Not sure how it works in Robolectric but if there is a TypedValue shim it should also use the conversion tables when appropriate.

**Describe alternatives you've considered**
None

**Additional context**
The Android source for 34 is not public yet, but you can find the same conversion tables in AndroidX: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-unit/src/androidMain/kotlin/androidx/compose/ui/unit/fontscaling/FontScaleConverterFactory.kt;l=24?q=fontscaleconverter&sq=

**Comment by hoisie** [comment]
cc @JuliaSullivanGoogle

**Comment by hoisie** [comment]
@odbol is this something that was introduced in Android 34, or is it possible to have non-linear font scaling with older SDKs?

With Robolectric Native Graphics enabled, you have real Android graphics code from Android 31 (Android S), which has Skia/Minikin, and supports native font scaling.

Is there some kind of basic test that can check for functionality of non-linear font scaling?

**Comment by odbol** [comment]
This was introduced in Android 34. Although Jetpack Compose will have it on older SDKs (not sure that matters to Robolectric). 

It is at a higher level than Skia: basically it happens during resource loading an XML dimension defined as SP, or when converting SP units to pixels using the TypedValue class. So I'm not sure if Robolectric picks up on that or not; it would have to use the real platform code for TypedValue in order to convert the fonts.

**Comment by sergio-sastre** [comment]
I can confirm it works out of the box with Robolectric 4.11-beta2, which support API 34.

use in your test:
```kotlin
@Config(
   sdk = [30, 33, 34],
   fontScale = 2.0f
)
```
And this is the result for the different SDKs:
| 30   | 33  | 34  |
| -------- | ------- | ------- |
|<img src="https://github.com/robolectric/robolectric/assets/6097181/863f71fb-3515-4040-83f0-3571fafceecf" width="150"/> | <img src="https://github.com/robolectric/robolectric/assets/6097181/3c2b948a-365c-46b1-9d5b-f43474b1a4b1" width="150"/> | <img src="https://github.com/robolectric/robolectric/assets/6097181/7b7ed684-0db5-4fe1-8e47-9ebf72643c38" width="150"/> |

**Comment by odbol** [comment]
Yep that looks like it's working. Very exciting! Thanks for testing 
-------- Original message --------From: Sergio Sastre Flórez ***@***.***> Date: 10/26/23  14:17  (GMT-08:00) To: robolectric/robolectric ***@***.***> Cc: Tyler Freeman ***@***.***>, Mention ***@***.***> Subject: Re: [robolectric/robolectric] Support non-linear font scaling (Issue #8413) 
I can confirm it works out of the box with Robolectric 4.11-beta2, which support API 34.
use in your test:
@Config(
   sdk = [30, 33, 34],
   fontScale = 2.0f
)
And this is the result for the different SDKs:



30
33
34
